### ISSUE TITLE
Slow convergence rate and lower accuracy than Adam?

### ISSUE THREAD

**Issue by githubharald** [XuezheMax/apollo#1]
Hi,

thanks for sharing your implementation.
I tried the Apollo optimizer for an image segmentation task and compared with the Adam optimizer.

The following plot shows a comparison (using validation metric "intersection over union" -> the larger the better) of Adam and Apollo using the same learning rate settings (1e-3). As you can see, Adam reaches ~84% after <10 epochs, while Apollo needs >200 epochs to reach ~82%. If changing the learning rate for Apollo, either the optimization diverges (for higher learning rates), or it takes even longer (for lower learning rates). Settings for Apollo: lr=1e-3, init_lr=1e-2, warmup=100.

![cmp](https://user-images.githubusercontent.com/15148095/95449167-ad12c480-0964-11eb-807b-608c23b91a9d.png)

Do you have any idea what's happening? 
How can I 1.) get faster convergence 2.) get higher accuracy than with Adam?

Best
Harald

**Comment by soloice** [comment]
Not the original author, but also interested in this paper.
I suppose `lr` should be larger than `init_lr`, which is for warm-up?
And based on my experience in training transformers, might you should lower down `init_lr` or use more `warmup` steps?

**Comment by githubharald** [comment]
you're right, warmup means that we should **start with a low lr**. I'll try again with init_lr=1e-5, lr=1e-3, warmup=1000 and will report the results.

**Comment by XuezheMax** [comment]
Thanks for your interests and thanks @soloice for the response.

First, I need to say that the range of hyper-parameter of Apollo is quite different from Adam and/or SGD. From the paper, you see that for Image classification the lr for Apollo is 0.5, and for NMT the lr is 10.0.
The reason why the lr ranges in a large area is that Apollo is **not** scale invariant (Adam is). For example, if we change our loss function from _L_ to _a x L_,  we need to adjust the lr of Apollo to _lr / a_ (For Adam, we need not adjust the lr). 
For the NMT task, we use token-level loss, which is pretty small. Thus, we need a much larger lr for Apollo on NMT.

I am not familiar with image segmentation, and not sure which architecture or loss you were using (CNN-based or attention-based?). Perhaps, you can provide more details, or paste the training log here?

For now, my suggestion is to start with lr=1.0, init_lr=0.01, and warmup=100. Note that the weight decay (if you used it) should be adjusted. For example, if you used lr=1e-3 and weight decay=1e-4 for Adam, you can try lr=1.0 and weight decay=1e-7 for Apollo. Another important hyper-parameter is gradient clip. It is important to clip the gradient for attention-based architecture (such as Transformer) to make the training stable.

**Comment by githubharald** [comment]
Thanks for the reply, now things start to look better!
Having 82% at epoch 5 with these settings: init_lr=1e-5, lr=0.1, warmup=100.
So the trick really was to start with a small lr and then finally use a quite large lr (should have read the Appendix of your paper too ;-) ).
Let's see how things look after 100-200 more epochs ... the goal would be to at least reach the ~90% that Adam achieves. I'll keep you updated.

**Comment by XuezheMax** [comment]
Thanks for your response!
Yes, the scale of lr for Apollo is much larger than Adam, and the warmup is super important for Apollo because of the initialization of the diagonal hessian approximation.

**Comment by githubharald** [comment]
Got 89.6%. Still slightly worse than Adam, but I guess with some more hyper-parameter tuning it might be possible to increase it even more.
I'll close this issue as I'm now in the accuracy range that I expected.

Update: having trained multiple models, I can now confirm that **Apollo is able to achieve the same accuracy as Adam** on our image classification task.
